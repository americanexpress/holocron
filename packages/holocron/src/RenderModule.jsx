/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { ReactReduxContext } from 'react-redux';
import PropTypes from 'prop-types';

import { getModule } from './moduleRegistry';

export default function RenderModule({ children, moduleName, props }) {
  const { store } = React.useContext(ReactReduxContext);

  const Module = getModule(moduleName, store.modules);
  if (!Module) {
    console.warn(`Module ${moduleName} was not found in the holocron module registry`);
    return null;
  }

  const propsWithChildren = typeof children !== 'undefined' ? { children, ...props } : props;

  // propsWithChildren is computed and passed down to this prop
  // eslint-disable-next-line react/jsx-props-no-spreading
  return <Module {...propsWithChildren} />;
}

RenderModule.propTypes = {
  moduleName: PropTypes.string.isRequired,
  // we're simply passing props through, the shape will be different for every module
  props: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  children: PropTypes.node,
};

RenderModule.defaultProps = {
  props: {},
  children: undefined,
};
