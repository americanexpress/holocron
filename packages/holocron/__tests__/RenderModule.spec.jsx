/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { ReactReduxContext } from 'react-redux';
// eslint-disable-next-line import/no-extraneous-dependencies
import { mount } from 'enzyme';
import { fromJS } from 'immutable';
import RenderModule from '../src/RenderModule';

// eslint-disable-next-line react/prop-types
const MyTestModule = ({ children, ...otherProperties }) => (
  <div>
    {JSON.stringify(otherProperties, undefined, 2)}
    {children}
  </div>
);

const store = {
  modules: fromJS({
    'my-test-module': MyTestModule,
  }),
  getState: jest.fn(() => fromJS({
    holocron: {
      loaded: {
        'my-test-module': true,
      },
    },
  })),
};

describe('RenderModule', () => {
  const consoleWarn = jest.spyOn(console, 'warn');

  beforeEach(() => jest.clearAllMocks());

  it('should warn and render null when it cannot find a module in registry', () => {
    expect.assertions(2);

    const tree = mount(
      <ReactReduxContext.Provider value={{ store }}>
        <RenderModule moduleName="not-in-module-map" />
      </ReactReduxContext.Provider>
    );
    expect(consoleWarn.mock.calls).toMatchSnapshot();
    expect(tree).toMatchSnapshot();
  });

  it('should warn and render null when module is not loaded', () => {
    store.getState.mockImplementationOnce(() => fromJS({
      holocron: {
        loaded: {},
      },
    }));
    expect.assertions(2);

    const tree = mount(
      <ReactReduxContext.Provider value={{ store }}>
        <RenderModule moduleName="not-in-module-map" />
      </ReactReduxContext.Provider>
    );
    expect(consoleWarn.mock.calls).toMatchSnapshot();
    expect(tree).toMatchSnapshot();
  });

  it('should render a module', () => {
    expect.assertions(1);

    const tree = mount(
      <ReactReduxContext.Provider value={{ store }}>
        <RenderModule moduleName="my-test-module" />
      </ReactReduxContext.Provider>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should pass props to the module', () => {
    expect.assertions(1);

    const properties = { hello: 'world', foo: 'bar' };
    const tree = mount(
      <ReactReduxContext.Provider value={{ store }}>
        <RenderModule moduleName="my-test-module" props={properties} />
      </ReactReduxContext.Provider>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should pass children to the module', () => {
    expect.assertions(1);

    const tree = mount(
      <ReactReduxContext.Provider value={{ store }}>
        <RenderModule moduleName="my-test-module"><h1>Hello, world</h1></RenderModule>
      </ReactReduxContext.Provider>
    );
    expect(tree).toMatchSnapshot();
  });
});
