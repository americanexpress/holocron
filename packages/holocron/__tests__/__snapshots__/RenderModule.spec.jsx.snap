// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderModule should pass children to the module 1`] = `
<RenderModule
  moduleName="my-test-module"
  props={Object {}}
>
  <MyTestModule>
    <div>
      {}
      <h1>
        Hello, world
      </h1>
    </div>
  </MyTestModule>
</RenderModule>
`;

exports[`RenderModule should pass props to the module 1`] = `
<RenderModule
  moduleName="my-test-module"
  props={
    Object {
      "foo": "bar",
      "hello": "world",
    }
  }
>
  <MyTestModule
    foo="bar"
    hello="world"
  >
    <div>
      {
  "hello": "world",
  "foo": "bar"
}
    </div>
  </MyTestModule>
</RenderModule>
`;

exports[`RenderModule should render a module 1`] = `
<RenderModule
  moduleName="my-test-module"
  props={Object {}}
>
  <MyTestModule>
    <div>
      {}
    </div>
  </MyTestModule>
</RenderModule>
`;

exports[`RenderModule should warn and render null when it cannot find a module in registry 1`] = `
Array [
  Array [
    "Module not-in-module-map was not found in the holocron module registry",
  ],
]
`;

exports[`RenderModule should warn and render null when it cannot find a module in registry 2`] = `
<RenderModule
  moduleName="not-in-module-map"
  props={Object {}}
/>
`;

exports[`RenderModule should warn and render null when module is not loaded 1`] = `
Array [
  Array [
    "Module not-in-module-map was not found in the holocron module registry",
  ],
]
`;

exports[`RenderModule should warn and render null when module is not loaded 2`] = `
<RenderModule
  moduleName="not-in-module-map"
  props={Object {}}
/>
`;
